### DISCUSSION TITLE
ES8388 vs AC101

### DISCUSSION THREAD

**Discussion by marcel-licence** [marcel-licence/esp32_midi_sampler#50] [General]
The latest version of this project does now also support the ES8388 (ref: https://youtu.be/8UB3fYPjqSk)
Please feel free to use this discussion for feedback, problems, ideas.

It looks like there are two unused IOs now. In addition to that it seems that mic and lineIn sharing one signal line.

**Comment by Phil-f6czv** [comment]
Hello
I found a specification of the new design with the ES8388
[esp32-a1s_v2.3_specification.pdf](https://github.com/marcel-licence/esp32_midi_sampler/files/7075076/esp32-a1s_v2.3_specification.pdf)

**Reply by gerthalberg** [reply]
Did you read page 6. I don't know if it's wishfull thinking on my part but it states: Integrate 64Mb PSRAM. Have you tried a check to see what it reports as available PSRAM?

**Reply by souravlal** [reply]
I also had a doubt that the ESP32 Audio-kit is 8MB, and asked about it earlier in this github page.
But when I select 8MB as storage in the settings and upload the display doesn't work.

**Reply by marcel-licence** [reply]
It looks like the 64Mb means 64Mbit -> 8 MB (byte) which could be an indicator of a configuration like the ESP32 wrover module. I found some discussion about accessing the complete memory. It looks like 4MB is usable easily. But to use the complete 8MB it will need a lot of effort.

**Reply by gerthalberg** [reply]
Apparently there is support for using the upper 4MB https://github.com/espressif/esp-idf/tree/master/examples/system/himem

**Reply by gerthalberg** [reply]
Btw. I now have 3 working boards. The original with the AC101, the one I thought was faulty and got a refund on and the one I ordered instead - with the refund :-)

**Reply by marcel-licence** [reply]
Hm it may be portable to arduino. I could give it a try soon. Thank you for the link. Good to hear that a refund might be possible.

**Reply by marcel-licence** [reply]
@souravlal Flashsize is only for your program (depends on your ESP32 variant). PSRAM is your RAM (4MB). There is no option to select the PSRAM size but it can be read by code because it's an external component connected to the mcu.

**Comment by schneedrache** [comment]
Hello!

Regarding the new A1S with the ES8388: Do you also have problems with the volume
when some headphones are directly connected to it?

Volume and quality is fine once the headphone output is connected to some
powered device (e.g. simple mixer without amp, my Korg NTS-1 unit, ...). Not so
when some headphones are directly connected to the headphone out of the A1s:
Much lower volume...

My guess is that it has something to do with the hardware. Same problem occurred
when I tried the mp3 example from the esp-adf repository...

**Reply by souravlal** [reply]
May be a simple preamp circuit required.

**Reply by Phil-f6czv** [reply]
I also noticed that the sound level is very weak when I flash an espressif example. I did not investigate.

**Reply by schneedrache** [reply]
@Phil-f6czv There seems to be a problem with the es8388 driver when using
examples from expressif. I guess that you have that one:

The registers for the _DACCONTROL26 (0x30)_ and _DACCONTROL27 (0x31)_ are set to
0 for some reason (maybe the LyraT uses the other two for headphone output...).

You need to set those to "**volume**" as done in the two lines above. The code
is located here: [GitHub: esp-adf: es8388.c](https://github.com/espressif/esp-adf/blob/0541882948f3732c2b10a47da35429f21bede2dc/components/audio_hal/driver/es8388/es8388.c#L356-L359)

The volume in the examples can then be set with audio_hal_set_volume. The value
70 worked fine for me:

```c
audio_hal_set_volume(board_handle->audio_hal, 70); // 0-100
```

This however does not solve my problem. Volume is still quiet with these
settings when using headphones directly...

**Reply by gerthalberg** [reply]
I have yet to start using the ES8388, I'm still using the the board with the AC - but I just scrolled through the ES code. Could it be related to  the es8388_pa_power(bool enable) or is it perhaps a gain related thing have you done anything with the input gain? either with es8388_set_adc_dac_volume or es8388_set_mic_gain. Have you tried to use the headphone_detect code even though it's not a Lyra board?

**Reply by marcel-licence** [reply]
There is one IO to turn on the amp's attached on the PCB. Marked here as AMP CTRL: https://youtu.be/8UB3fYPjqSk?t=165

**Reply by gerthalberg** [reply]
On a LyraT 4.2/4.2 it has the following defined #define PA_ENABLE_GPIO            GPIO_NUM_21. This corresponds nicely to the es8388_pa_power(bool enable) that needs to know that exact pin, to enable pa_power i.e. turn on the amplifiers. Furthermore - watch the following YT. He mentions the exact fix above for Register 26 and 27 https://www.youtube.com/watch?v=q5wghoIwrkg.

**Reply by schneedrache** [reply]
Setting the Pin 21 (AMP CTRL...) has no effects for me. It seems like that only the speaker outputs are affected (the ones with the NS4150), but I currently have no connector available for those pins so I cannot test it.

I will look further into possible problem sources. Otherwise the setup with my mixer or with my audio interface is fine for now, just not very portable.

**Reply by gerthalberg** [reply]
How about a couple of dupont connectors. If you remove the connector from the housing it'll work - otherwise they are to long.

**Reply by schneedrache** [reply]
~Those type of connectors are not the problem,rather it is that I can't connect them to something (e.g. TRS 3.5mm breakout -> headphones, directly to speakers...). I also think that connecting normal _headphones_ to the (3-watt?) _speaker_ output, especially when the amp is enabled, is not a good idea.~

~Maybe I will check that with my multimeter, so I can see that setting the pin has an effect here. But I still think that this is unrelated to the problem with the headphone jack, though I might be wrong there...~

**EDIT:** Found old headphones and a board with a jack from some old phone and soldered some wires to it. The amp setting works and I get some sound from it. Volume is also OK, but the output is very noisy. Also, there is no output at all if the amp is deactivated via the pin 21.

However, the setting does not effect the headphone jack. I also have the feeling that the low end of the sound on the headphone jack is not really there at all, but this might just be due to the low volume...

**Reply by gerthalberg** [reply]
What about the gain settings. I have had success whit that on other boards and codecs that were weak on the output. That would fit with missing low end volume and high noise level.

**Reply by schneedrache** [reply]
I played around both with the volume register of the ES8388 as well as with the variable "master_output_gain". I actually got some acceptable volume levels, but the low end is really not that present (only the mids and highs are present like in some cheap radio or so...).

However, the sound quality and the volume is still way better when I connect the A1S to my mixer, audio interface, effects unit. I have no amplification enabled on all of these devices. That is also the reason why I think that it has something to do with the hardware. But my knowledge about the hardware stuff is really limited.

My next guess is that there is a power problem with the output. But I also tested it with my 16 Ω IEMs and some really cheap gaming headphones. Same result...

**Reply by marcel-licence** [reply]
The line level on the ES8388 seems to be okay. I created some distorted sounds. In that case the max volume seems to be the same as the AC101. Using a headphone I observed also lower output with quieter low frequencies.
With respect to the datasheet. The output supports 40mW headphone. (https://datasheet.lcsc.com/szlcsc/1901241705_Everest-Semiconductor-ES8388_C365736.pdf)
Maybe some low power mode is still active.

**Reply by gerthalberg** [reply]
/* Set ALC and Gain */
    ES8388_WriteReg(ES8388_ADCCONTROL10, 0xFF); // 0xFF: ALC = Stereo, MAXGAIN = 35.5Db MINGAIN = 30Db

This setting works wonder for the output level on the headphones from barely audible to blowing your head off!

**Comment by gerthalberg** [comment]
Link to ES8388 User Guide without the Chinese yellow stickers

https://manualzz.com/doc/o/mzzhv/es8388-user-guide-fully-differential-microphone-input-circuit-and-sample-code

**Comment by GillesLACAUD** [comment]
Hi Marcel

I have buy some other ESP32 Audio Kit boards and of course now i have the ES8388 issue.
So i have change the I2C pin to 18 and 23 and change the I2C address.
In fact i use your ES8388 code.
But my init ES8388 return always fail.

Did i miss something ?
There is an hardware modification ?

Thank again for you help

**Reply by gerthalberg** [reply]
I Have notice something a little strange: When using the Arduino Wire library I have to use 0x10 as the chip address however when using ESP IDF and the I2C library it appears, that I have to use 0x20 as the chip address. The physical board used is the same.

**Reply by GillesLACAUD** [reply]
Thx but i think it is only à logical shift with the i2c address. With à addr at 0x10 i have à nack. So i think thé réal addr is maybe 0x11..somdthing wrong around thé ad0 and thé pull up and pull down resistor on the codec
![Uploading DSC_1693.JPG…]()

**Reply by hamuro80** [reply]
In my case with blackstomp board, I have to use 0x10 as the address... it is right according to the es8388 datasheet, but it is wrong according to es8388 user guide document.

**Reply by GillesLACAUD** [reply]
It is ok in the 2 cases. In fact it depends how the driver manages the I2C address.
for the ES8388 the i2c address is **0010 000x** where x is the Read or Write bit. So you can said it is an **0x20** address (but 0x21 will work too).
On the other side you can said the address is 0x10 if you delete the R/W bit and in this case the driver will apply a logical left shift
**0x10<<1 = 0x20** and then add the R/W bit.

**Reply by GillesLACAUD** [reply]
Did you have some sound in the audio Jack ?

**Comment by GillesLACAUD** [comment]
I have find something interresting.
When i scan the I2C bus pin 33/32 i find a device at the 0x10 address.
When i scan the I2C bus pin 18/23 i find a no device.

If i init the ES8388 with the I2C 33/32 at address 0x10 -> **THE ES8388 INIT IS OK**

So i think they have change the routing for the I2C ES8388 in the ESP32-AS1....

**Comment by GillesLACAUD** [comment]
Some news from AI....configuration of the I2C and I2S for the latest boards.
Not tested on my side yet
![image](https://user-images.githubusercontent.com/10655745/138229949-50d1a3a4-365a-4618-ac56-71c126649065.png)

**Comment by GillesLACAUD** [comment]
it was a little bit long....but now i think i have the good information
[https://github.com/Ai-Thinker-Open/ESP32-A1S-AudioKit/issues/26
My only doubt now is about the in and out pins. Is datain_num pin35 view from the ESP32 side or the ES8388 side (same for the out of course)

**So be careful if you buy another board soon....
_We already to halt producte the an ES8388 module with 18/23 on I2C**_

**Comment by GillesLACAUD** [comment]
The data_in_num pin n°35 is view from the ESP32 side, because the pin n°35 is an only input pin.
